modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java

```diff
@@ -17,60 +17,310 @@
+                        else if (evt.newEntry() == null) // Table Dropped.
+                            schemes.remove(tblId);
+                        else //TODO: https://issues.apache.org/jira/browse/IGNITE-13752
+                            throw new SchemaRegistryException("Schema upgrade is not implemented yet.");
+                    }
+                    else {
+                        UUID tblId = UUID.fromString(keyTail.substring(0, verPos));
+
+                        final SchemaRegistryImpl reg = schemes.get(tblId);
+
+                        assert reg != null : "Table schema was not initialized or table has been dropped: " + tblId;
+
+                        reg.registerSchema((SchemaDescriptor)ByteUtils.fromBytes(evt.newEntry().value()));
+                    }
+                }
+
+                return true;
             }
+
+            /** {@inheritDoc} */
+            @Override public void onError(@NotNull Throwable e) {
+                LOG.error("Metastorage listener issue", e);
+            }
+        });
+    }
+
+    /**
+     * Unsubscribes a listener form the affinity calculation.
+     */
+    private void unsubscribeFromAssignmentCalculation() {
+        if (schemaHistorySubscriptionFut == null)
+            return;
+
+        try {
+            Long subscriptionId = schemaHistorySubscriptionFut.get();
+
+            metaStorageMgr.unregisterWatch(subscriptionId);
+
+            schemaHistorySubscriptionFut = null;
+        }
+        catch (InterruptedException | ExecutionException e) {
+            LOG.error("Couldn't unsubscribe for Metastorage updates", e);
+        }
+    }
+
+    /**
+     * Reads current schema configuration, build schema descriptor,
+     * then add it to history rise up table schema version.
+     *
+     * @param tblId Table id.
+     * @param tblName Table name.
+     * @return Operation future.
+     */
+    public CompletableFuture<Boolean> initNewSchemaForTable(UUID tblId, String tblName) {
+        return vaultManager.get(ByteArray.fromString(INTERNAL_PREFIX + tblId)).
+            thenCompose(entry -> {
+                TableConfiguration tblConfig = configurationMgr.configurationRegistry().getConfiguration(TablesConfiguration.KEY).tables().get(tblName);
+                var key = new Key(INTERNAL_PREFIX + tblId);
+
+                int ver = entry.empty() ? 1 : (int)ByteUtils.bytesToLong(entry.value(), 0) + 1;
+
+                final SchemaDescriptor desc = createSchemaDescriptor(tblConfig, ver);
+
+                return metaStorageMgr.invoke(
+                    Conditions.key(key).value().eq(entry.value()), // Won't to rewrite if the version goes ahead.
+                    List.of(
+                        Operations.put(key, ByteUtils.longToBytes(ver)),
+                        Operations.put(new Key(INTERNAL_PREFIX + tblId + INTERNAL_VER_SUFFIX + ver), ByteUtils.toBytes(desc))
+                    ),
+                    List.of(
+                        Operations.noop(),
```

Comment:
One Operations.noop() is enough, it's not necessary to have same amount of failure operations as success ones.